### PR TITLE
feat(assistant): Alert Rules Guide

### PR DESCRIPTION
**Pull request by ehfeng**

### CODE REVIEW

File: src/sentry/assistant/guides.py [hunk 1]
```diff
@@ -150,5 +152,54 @@
...
+            #     'title': _('Ownership'),
+            #     'message': _(),
+            #     'target': None
+            # },
+            {
+                'title': _('Tweaking personal email settings'),
+                'message': _('If you\'d like to customize your <i>personal</i> alert settings, '
+                             'go to <a href="/account/settings/notifications/">Account Notifications</a>. '
+                             'There, you can choose which project\'s alert or workflow notifications you\'d like to receive.'),
+                'target': None,
```
Comment (adhiraj) [inline src/sentry/assistant/guides.py]: Can you add a comment to the top of the file about how targets behave. it's getting confusing
- target: 'non-empty-string' = step is tied to an anchor target. if the anchor doesn't exist, the step will not be shown. if the anchor exists but is of type "invisible", it will not be pinged but will be scrolled to. otherwise the anchor will be pinged and scrolled to.
- target: None = step isn't tied to any particular element, and will never be removed from the guide.

if this gets too confusing I might make a pass to clean things up

File: src/sentry/static/sentry/app/views/settings/projectAlerts/projectAlertRules.jsx [hunk 2]
```diff
@@ -212,6 +226,7 @@ class ProjectAlertRules extends AsyncView {
               params={this.props.params}
               routes={this.props.routes}
               onDelete={this.handleDeleteRule.bind(this, rule)}
+              firstRule={this.state.ruleList.indexOf(rule) === 0}
```
Comment (adhiraj) [inline src/sentry/static/sentry/app/views/settings/projectAlerts/projectAlertRules.jsx]: More idiomatic to use the indexed form of `map` to get the index. See https://github.com/getsentry/sentry/blob/master/src/sentry/static/sentry/app/utils/stream.jsx#L50

File: src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx [hunk 3]
```diff
@@ -79,6 +79,9 @@ const GuideAnchor = createReactClass({
   },
 });
 
+export const ConditionalGuideAnchor = ({condition, target, type, children}) =>
```
Comment (adhiraj) [inline src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx]: What's the benefit of this over the (I think simpler) `condition && <GuideAnchor ...>`?

Comment (ehfeng) [reply src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx]: The benefits are definitely more hypothetical. The primary reason I'd done the wrapping was a future world where we might put the pings *behind* the UI element, so the user can still click on the element itself mid-tour.

It's obviously not a vital feature, but I don't see much downside to `ConditionalGuideAnchor` except perhaps, we'd like to fold this into the `GuideAnchor` class itself.

File: src/sentry/assistant/guides.py (hunk 1 above)
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: @denamwangi, I noticed you used "invisible" anchors, which might be the better solution. Thoughts?

File: src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx (hunk 3 above)
Comment (ehfeng) [reply src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx]: The advantages of wrapping (versus inserting) is that we have greater control of the actual element we're trying to focus and (for instance) can capture and pass down click events (whereas adjacent elements aren't aware of the target element).

We're not leveraging this now, but it feels better as a pattern.

File: src/sentry/assistant/guides.py [hunk 4]
```diff
@@ -150,5 +152,54 @@
                 'target': 'member_role',
             },
         ]
-    }
+    },
+    # Ideally, this would only be sent if the organization has never
+    # customized alert rules (as per FeatureAdoption)
+    'alert-rules': {
```
[the remark this answers is not in the document]
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: done

File: src/sentry/assistant/guides.py [hunk 5]
```diff
@@ -150,5 +152,54 @@
...
+        'id': 5,
+        'cue': _('Tips for alert rules'),
+        'required_targets': ['alert_conditions'],
+        'steps': [
+            {
+                'title': _('Reducing alert noise'),
+                'message': _('Sentry, by default, alerts on every <i>new</i> issue via email. '
```
[the remark this answers is not in the document]
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: done

File: src/sentry/assistant/guides.py [hunk 6]
```diff
@@ -150,5 +152,54 @@
...
+                             'If that\'s too noisy, sending the new issue alerts to '
+                             'a service like Slack help reduce inbox noise.'),
+                'target': 'alert_conditions',
+            },
+            {
+                'title': _('Defining priority alerts'),
+                'message': _('Not alerts are equal. Create rules for frequently occuring errors or specific '
```
[the remark this answers is not in the document]
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: done

File: src/sentry/assistant/guides.py [hunk 7]
```diff
@@ -150,5 +152,54 @@
...
+                             'tags to escalate critical issues by alerting you via email or PagerDuty.'),
+                'target': 'alert_actions',
+            },
+            # When ownership is released
```
[the remark this answers is not in the document]
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: done

File: src/sentry/assistant/guides.py [hunk 8]
```diff
@@ -150,5 +152,54 @@
...
+    },
+    # Ideally, this would only be sent if the organization has never
+    # customized alert rules (as per FeatureAdoption)
+    'alert-rules': {
+        'id': 5,
+        'cue': _('Tips for alert rules'),
+        'required_targets': ['alert_conditions'],
+        'steps': [
+            {
+                'title': _('Reducing alert noise'),
```
[the remark this answers is not in the document]
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: done

File: src/sentry/assistant/guides.py [hunk 9]
```diff
@@ -150,5 +152,54 @@
...
+            },
+            # When ownership is released
+            # {
+            #     'title': _('Ownership'),
+            #     'message': _(),
+            #     'target': None
+            # },
+            {
+                'title': _('Tweaking personal email settings'),
+                'message': _('If you\'d like to customize your <i>personal</i> alert settings, '
```
[the remark this answers is not in the document]
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: done

File: src/sentry/assistant/guides.py [hunk 10]
```diff
@@ -150,5 +152,54 @@
...
+                             'go to <a href="/account/settings/notifications/">Account Notifications</a>. '
+                             'There, you can choose which project\'s alert or workflow notifications you\'d like to receive.'),
+                'target': None,
+            },
+            {
+                'title': _('What are legacy integrations?'),
```
[the remark this answers is not in the document]
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: I like the question better, just because I like the "oh, and this thing too" casual tone it has.

File: src/sentry/assistant/guides.py (hunk 1 above)
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: documented a bit better, but still need to decide what best behavior is here.

File: src/sentry/static/sentry/app/views/settings/projectAlerts/projectAlertRules.jsx (hunk 2 above)
Comment (ehfeng) [reply src/sentry/static/sentry/app/views/settings/projectAlerts/projectAlertRules.jsx]: done

**Comment by ehfeng** [comment]
rebasing because of Ted's snuba commits are breaking my dev env.

**Comment by ehfeng** [comment]
Rebasing because of the new js import pattern.

File: src/sentry/assistant/guides.py [hunk 11]
```diff
@@ -3,6 +3,24 @@
 from django.utils.translation import ugettext_lazy as _
 
 
+# Guide Schema
+# id (text, required): unique id
+# cue (text, optional):  The text used to prmopt the user to initiate the guide.
+#                        Leaving this blank
```
Comment (adhiraj) [inline src/sentry/assistant/guides.py]: ?

File: src/sentry/assistant/guides.py [hunk 12]
```diff
@@ -3,6 +3,24 @@
...
+# required_targets (list): An empty list will cause the guide to be shown regardless
```
Comment (adhiraj) [inline src/sentry/assistant/guides.py]: ?

File: src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx [hunk 13]
```diff
@@ -20,21 +20,31 @@ const GuideAnchor = createReactClass({
     target: PropTypes.string.isRequired,
     // The `invisible` anchor type can be used for guides not attached to specific elements.
     type: PropTypes.oneOf(['text', 'button', 'invisible']),
+    show: PropTypes.bool,
```
Comment (adhiraj) [inline src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx]: undo these changes

File: src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx [hunk 14]
```diff
@@ -79,6 +94,12 @@ const GuideAnchor = createReactClass({
   },
 });
 
+export const conditionallyGuideAnchor = (condition, target, type, children) => {
```
Comment (adhiraj) [inline src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx]: conditionally sounds weird. rename to `conditionGuideAnchor`?

Comment (ehfeng) [reply src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx]: sure

File: src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx (hunk 13 above)
Comment (ehfeng) [reply src/sentry/static/sentry/app/components/assistant/guideAnchor.jsx]: done

File: src/sentry/assistant/guides.py (hunk 12 above)
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: done

File: src/sentry/assistant/guides.py (hunk 11 above)
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: i think we were discussing something, but i can't remember the utility of empty cues.

File: src/sentry/assistant/guides.py [hunk 15]
```diff
@@ -3,6 +3,24 @@
 from django.utils.translation import ugettext_lazy as _
 
 
+# Guide Schema
+# id (text, required): unique id
+# cue (text, optional):  The text used to prompt the user to initiate the guide.
```
Comment (adhiraj) [inline src/sentry/assistant/guides.py]: I don't think any of the fields we have can be optional? (in the sense you may just omit it instead of making it [] or None)

File: src/sentry/assistant/guides.py [hunk 16]
```diff
@@ -150,5 +170,54 @@
...
+        'id': 5,
+        'cue': _('Tips for alert rules'),
+        'required_targets': ['alert_conditions'],
+        'steps': [
+            {
+                'title': _('Reduce inbox noise'),
+                'message': _('Sentry, by default, alerts on every <i>new</i> issue via email. '
+                             'If that\'s too noisy, sending the new issue alerts to '
+                             'a service like Slack help reduce inbox noise.<br><br> Enabling '
+                             '<a href="https://sentry.io/settings/account/notifications/#weeklyReports">'
```
Comment (adhiraj) [inline src/sentry/assistant/guides.py]: add target="_blank" to all links (we don't want them to navigate away from the page in the middle of a tour)

File: src/sentry/assistant/guides.py [hunk 17]
```diff
@@ -150,5 +170,54 @@
...
+                             'If you want an alert rule to trigger both legacy and global '
+                             'integrations, you need to add both as actions. '
+                             '<a href="https://help.sentry.io/hc/en-us/articles/360003063454">Learn more</a>.'),
+                'target': None,
+            },
+        ],
+    },
 }
+
+# Configure -> next steps???
```
Comment (adhiraj) [inline src/sentry/assistant/guides.py]: what are these

File: tests/js/spec/components/__snapshots__/streamGroup.spec.jsx.snap [hunk 18]
```diff
@@ -11,6 +11,7 @@ exports[`StreamGroup renders with anchors 1`] = `
     ml={2}
   >
     <GuideAnchor
+      show={true}
```
Comment (adhiraj) [inline tests/js/spec/components/__snapshots__/streamGroup.spec.jsx.snap]: update specs

File: src/sentry/assistant/guides.py (hunk 15 above)
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: Ah, that's true. I'll fix.

File: src/sentry/assistant/guides.py (hunk 16 above)
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: Good point.

File: tests/js/spec/components/__snapshots__/streamGroup.spec.jsx.snap (hunk 18 above)
Comment (ehfeng) [reply tests/js/spec/components/__snapshots__/streamGroup.spec.jsx.snap]: ugh, forgot to stage them.

File: src/sentry/assistant/guides.py (hunk 17 above)
Comment (ehfeng) [reply src/sentry/assistant/guides.py]: removed. cruft from development.

**Review by adhiraj — APPROVED** [review]
yeah!
